protocol/common/raw_messages_persistence.go

```diff
@@ -336,3 +337,78 @@ func (db RawMessagesPersistence) DeleteHashRatchetMessages(ids [][]byte) error {
 
 	return err
 }
+
+func (db *RawMessagesPersistence) IsMessageAlreadyCompleted(hash []byte) (bool, error) {
+	var alreadyCompleted int
+	err := db.db.QueryRow("SELECT COUNT(*) FROM message_segments_completed WHERE hash = ?", hash).Scan(&alreadyCompleted)
+	if err != nil {
+		return false, err
+	}
+	return alreadyCompleted > 0, nil
+}
+
+func (db *RawMessagesPersistence) SaveMessageSegment(segment *protobuf.SegmentMessage, sigPubKey *ecdsa.PublicKey) error {
+	sigPubKeyBlob := crypto.CompressPubkey(sigPubKey)
+
+	_, err := db.db.Exec("INSERT INTO message_segments (hash, segment_index, segments_count, sig_pub_key, payload) VALUES (?, ?, ?, ?, ?)",
+		segment.EntireMessageHash, segment.Index, segment.SegmentsCount, sigPubKeyBlob, segment.Payload)
+
+	return err
+}
+
+// Get ordered message segments for given hash
+func (db *RawMessagesPersistence) GetMessageSegments(hash []byte, sigPubKey *ecdsa.PublicKey) ([]*protobuf.SegmentMessage, error) {
+	sigPubKeyBlob := crypto.CompressPubkey(sigPubKey)
+
+	rows, err := db.db.Query("SELECT hash, segment_index, segments_count, payload FROM message_segments WHERE hash = ? AND sig_pub_key = ? ORDER BY segment_index", hash, sigPubKeyBlob)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var segments []*protobuf.SegmentMessage
+	for rows.Next() {
+		var segment protobuf.SegmentMessage
+		err := rows.Scan(&segment.EntireMessageHash, &segment.Index, &segment.SegmentsCount, &segment.Payload)
+		if err != nil {
+			return nil, err
+		}
+		segments = append(segments, &segment)
+	}
+	err = rows.Err()
+	if err != nil {
+		return nil, err
+	}
+
+	return segments, nil
+}
+
+func (db *RawMessagesPersistence) CompleteMessageSegments(hash []byte, sigPubKey *ecdsa.PublicKey) error {
+	tx, err := db.db.BeginTx(context.Background(), &sql.TxOptions{})
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if err == nil {
+			err = tx.Commit()
+			return
+		}
+		// don't shadow original error
+		_ = tx.Rollback()
+	}()
+
+	sigPubKeyBlob := crypto.CompressPubkey(sigPubKey)
+
+	_, err = tx.Exec("DELETE FROM message_segments WHERE hash = ? AND sig_pub_key = ?", hash, sigPubKeyBlob)
+	if err != nil {
+		return err
+	}
+
+	_, err = tx.Exec("INSERT INTO message_segments_completed (hash, sig_pub_key) VALUES (?,?)", hash, sigPubKeyBlob)
```

Comment:
Do you think we'd need to clean up that table at some point? After 30 days, the entries here are pretty useless.
They should be small, so it's not a big issue, but it does take some space and the bigger the table the longer queries will take.
It can be done in another PR too

Reply:
Yes, that would make sense :+1:  I planned to create a similar follow-up task to clean up the `message_segments` table in case segments aren't completed in X days/hours, because it would occupy even more memory, since the message segments themselves are much heavier than a tuple of two hashes. I'll make sure that task includes your proposal as well.

Reply:
@jrainville FYI: https://github.com/status-im/status-go/issues/4297